app/assets/javascripts/wizard.js

```diff
@@ -0,0 +1,113 @@
+(function ($) {
+
+  var WIZARD_PLUGIN_NS = 'refusalWizard';
+
+  var Plugin = function (target, options) {
+    this.$el = $(target);
+
+    var defaults = {
+      questionClass: 'wizard__question',
+      currentQuestionClass: 'wizard__question--current',
+      pastQuestionClass: 'wizard__question--past',
+      futureQuestionClass: 'wizard__question--future',
+      remainingMessageClass: 'js-wizard-remaining',
+      progressBarClass: 'js-wizard-progress',
+      nextButtonClass: 'js-wizard-next'
+    };
+
+    this.options = $.extend(true, defaults, options);
+
+    this._init(target, options);
+
+    return this;
+  };
+
+  Plugin.prototype._init = function (target, options) {
+    var _this = this;
+    this._updateTotals();
+    this._updateUI();
+
+    $('.' + this.options.nextButtonClass).on('click', function () {
+      _this._submitAnswer();
+    });
+  };
+
+  Plugin.prototype._updateTotals = function () {
+    this.totalQuestions = this.$el.find('.' + this.options.questionClass).length;
+    this.totalAnswered = this.$el.find('.' + this.options.pastQuestionClass).length;
+    this.totalUnanswered = this.totalQuestions - this.totalAnswered;
+  };
+
+  Plugin.prototype._updateUI = function () {
+    console.log(this);
+    var remainingString = '' + this.totalUnanswered + ' question' + (this.totalUnanswered === 1 ? '' : 's') + ' remaining';
+    var progressPercent = ( this.totalQuestions - this.totalUnanswered ) / this.totalQuestions * 100;
+
+    this.$el.find('.' + this.options.remainingMessageClass).text(remainingString);
+    this.$el.find('.' + this.options.progressBarClass).css({
+      'width': '' + progressPercent + '%'
+    }).attr({
+      'aria-valuenow': this.totalQuestions - this.totalUnanswered,
+      'aria-valuemax': this.totalQuestions
+    });
+  };
+
+  Plugin.prototype._submitAnswer = function() {
+    var $current = this.$el.find('.' + this.options.currentQuestionClass);
+    var $next = $current.next('.' + this.options.questionClass);
+
+    $current.removeClass(this.options.currentQuestionClass);
+    $current.addClass(this.options.pastQuestionClass);
+    $next.removeClass(this.options.futureQuestionClass);
+    $next.addClass(this.options.currentQuestionClass);
+
+    this._updateTotals();
+    this._updateUI();
+  };
+
+  // Declare public methods here. They should return the
+  // plugin target (this.$el) for jQuery chaining, eg:
+  //
+  // Plugin.prototype.examplePublicMethod = function () {
+  //   return this.$el;
+  // };
+
+  $.fn[ WIZARD_PLUGIN_NS ] = function( methodOrOptions ) {
```

Comment:
The code from this point took me a while to understand what it was doing. Am I correct in assuming this is boilerplate code for custom jQuery plugins to get the `.refusalWizard` method on jQuery elements?

I'm curious why you chose a plugin as I don't think we've currently used this approach anywhere else in Alaveteli?

I've tired stayed away from using jQuery inorder to make life easier when upgrading to webpack once we get the Rails upgrades out of the way.

It a shame thats not ready yet now that we're adding more complex UI elements. For now I'm happy to continue with this jQuery approach.

Reply:
I started writing it as just a collection of lines of javascript, but once I broke different parts of the process out into separate functions, I started spending a lot of time passing references to DOM elements around between those functions, and setting global variables to hold information about the wizard’s state.

Wrapping it all up as a jQuery plugin solved a lot of these issues for me, and helped me keep the actual nuts and bolts of the interaction stuff quite simple. But I agree, that comes at the cost of quite a lot of boilerplate.

Happy to pair on transforming it to a different format, if you’d prefer?